source/inputoutput.rst

```diff
@@ -3,15 +3,113 @@ Inputs and Outputs in BharatSim
-* data gets 'ingested' and graphNodes are created
+BharatSim uses a `CSV <https://en.wikipedia.org/wiki/Comma-separated_values>`_ file as an input. It is equipped to **ingest** data from a file, by reading it and converting the data to the network.
+
+In order to ingest data, we need to make use of three functions:
+
+* ``ingestData``, which is a method of the ``simulation`` class
+* ``ingestCSVData``, a method of the ``ContextBuilder`` object
+* A user-defined function which tells the program what data to extract and what to do with it
+
+Using the framework-defined functions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+First, we need to import the necessary packages:
+
+.. code-block:: scala
+
+  import com.bharatsim.engine.ContextBuilder._
+  import com.bharatsim.engine.execution.Simulation
+  import com.bharatsim.engine.graph.ingestion.{GraphData, Relation}
+
+.. hint:: You'll see why we import ``GraphData`` and ``Relation`` in the next section!
+
+The next step is to create an instance of the simulation class in the main function,
+
+.. code-block:: scala
+
+  val simulation = Simulation()
+
+We then ingest the data in the following way:
+
+.. code-block:: scala
+
+  simulation.ingestData(implicit context => {
+    ingestCSVData("citizen10k.csv", csvDataExtractor)
+    logger.debug("Ingestion done")
+  })
+
+where ``csvDataExtractor`` is the user-defined function.
+
+.. note:: The above block of code essentially causes the data from the CSV file to be read one line at a time
+
+Using the User-Defined function
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The user-defined function (``csvDataExtractor``, in our case) will depend on the data we want to extract. As an example, let's consider that we have data on a number of cats, each with their own ID, name, city of residence, an integer ID for the city, and a particular colour. Our CSV file would look like
+
+.. code-block:: csv
+
+  ID,Name,City,CityID,Colour
+  0,Mittens,Sydney,2000,White
+  1,Tabby,London,1050,Brown
+  2,Garfield,LasagnaLand,7,Orange
+  3,Elizabeth,Mishelam,102,Black
+  4,Coppe,Crossbell,100,Black
+  5,Antoine,Zeiss,62,Brown
+
+Let's assume we've already defined the following:
+
+* a case class ``Cat`` with three attributes, ``id``, ``name`` and ``colour``
+* another case class ``City`` with two attributes, ``id`` and ``cityname``
+We could then write out ``csvDataExtractor`` function in the following way:
+
+.. code-block:: scala
+
+  private def csvDataExtractor(map: Map[String, String](implicit context: Context): GraphData = {
+
+    val catName = map("Name").toString
+    val catID = map("ID").toLong
+    val catCity = map("City").toString
+    val catCityID = map("CityID").toLong
+    val catColour = map("Colour").toString
+
+    val singleCat: Cat = Cat(
+      catID,
+      catName,
+      catColour
+    )
+
+    val singleCity: City = City(
+      catCityId,
+      catCity
+    )
+
+    val staysAt = Relation[Cat, City](catID, "STAYS_AT", catCityID)
```

Comment:
Shouldn't you have two relations here? One between the ``Cat`` and the ``City``, and one between the ``City`` and the ``Cat``? Check the examples in the code to see what I mean.

Reply:
Is that necessary? I figured that just 1 relation would be enough to keep it simple

I can still add it though, it's pretty simple

Reply:
It's not only necessary, it's essential. Without both relations, the framework won't work the way its supposed to, if I remember correctly.

Reply:
My bad then, I didn't know that.

I'll add it and look into why